SimulationAgent/Simulation/DeviceStatusLogic/UpdateReportedProperties.cs

```diff
@@ -0,0 +1,172 @@
+                    "Checking for desired property updates & updated reported properties",
+                    () => new { this.deviceId, deviceState = actor.DeviceState
+                    });
+
+                // Get device
+                var device = this.GetDevice(actor.CancellationToken);
+                lock (actor.DeviceState)
+                {
+                    // check for differences between reported/desired properties,
+                    // update reported properties with any state changes (either from desired prop changes, methods, etc.)
+                    if (this.ChangeTwinPropertiesToMatchDesired(device, actor.DeviceState)
+                        || this.ChangeTwinPropertiesToMatchActorState(device, actor.DeviceState))
+                        actor.BootstrapClient.UpdateTwinAsync(device).Wait((int)connectionTimeout.TotalMilliseconds);
+                }
+
+
+                // Move state machine forward to start sending telemetry messages if needed
+                if (actor.ActorStatus == Status.UpdatingReportedProperties)
+                {
+                    actor.MoveNext();
+                }
+                else
+                {
+                    this.log.Debug(
+                        "Already moved state machine forward, continuing to check for desired property changes",
+                        () => new { this.deviceId });
+                }
+
+            }
+            catch (Exception e)
+            {
+                this.log.Error("UpdateReportedProperties failed",
+                    () => new { this.deviceId, e.Message, Error = e.GetType().FullName});
+            }
+            finally
+            {
+                //TODO: Here we should unpause the timer - this same thing should be done in all state machine methods
+
+            }
+
+        }
+
+
+        public void Setup(string deviceId, DeviceModel deviceModel)
+        {
+
+            if (this.setupDone)
+            {
+                this.log.Error("Setup has already been invoked, are you sharing this instance with multiple devices?",
+                    () => new { this.deviceId });
+                throw new DeviceActorAlreadyInitializedException();
+            }
+
+            this.setupDone = true;
+            this.deviceId = deviceId;
+            this.deviceModel = deviceModel;
+
+        }
+
+        private bool ChangeTwinPropertiesToMatchActorState(Device device, Dictionary<string, object> actorState)
+        {
+            bool differences = false;
+
+            foreach (var item in actorState)
+            {
+                if (device.Twin.ReportedProperties.ContainsKey(item.Key))
+                {
+                    if (device.Twin.ReportedProperties[item.Key].ToString() != actorState[item.Key].ToString())
+                    {
+                        // Update the Hub twin to match the actor state
+                        device.Twin.ReportedProperties[item.Key] = actorState[item.Key].ToString();
```

Comment:
check whether this transforms a Number to a String, e.g. see latitude/longitude

Reply:
this functions; i.e. it doesn't change the type, leaving the number as a number.